File: DsWorkJava/dswork.http/dswork/http/HttpUtil.java
package dswork.http;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLSocketFactory;

/**
 * 封装http请求
 * @author skey
 * @version 1.0
 */
public class HttpUtil
{
	private HttpURLConnection http;
	private SSLSocketFactory sslSocketFactory;
	private boolean isHttps = false;
	private int connectTimeout = 10000;
	private int readTimeout = 30000;
	private String userAgent = "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.104";

	/**
	 * 返回当前是否https请求
	 * @return boolean
	 */
	public boolean isHttps()
	{
		return isHttps;
	}

	/**
	 * 设置超时时间毫秒
	 * @param connectTimeout int
	 * @return HttpUtil
	 */
	public HttpUtil setConnectTimeout(int connectTimeout)
	{
		this.connectTimeout = connectTimeout;
		if(http != null)
		{
			this.http.setConnectTimeout(connectTimeout);
		}
		return this;
	}

	/**
	 * 设置contentType
	 * @param contentType String
	 * @return HttpUtil
	 */
	public HttpUtil setContentType(String contentType)
	{
		if(http != null)
		{
			this.http.setRequestProperty("Content-Type", contentType);
		}
		return this;
	}

	/**
	 * 设置读取超时时间毫秒
	 * @param readTimeout int
	 * @return HttpUtil
	 */
	public HttpUtil setReadTimeout(int readTimeout)
	{
		this.readTimeout = readTimeout;
		if(http != null)
		{
			this.http.setReadTimeout(readTimeout);
		}
		return this;
	}

	/**
	 * 设置requestMethod
	 * @param requestMethod String
	 * @return HttpUtil
	 */
	public HttpUtil setRequestMethod(String requestMethod)
	{
		try
		{
			this.http.setRequestMethod(requestMethod.toUpperCase(Locale.ROOT));
		}
		catch(Exception e)
		{
		}
		return this;
	}

	/**
	 * 设置requestProperty
	 * @param key String
	 * @param value String
	 * @return HttpUtil
	 */
	public HttpUtil setRequestProperty(String key, String value)
	{
		try
		{
			this.http.setRequestProperty(key, value);
		}
		catch(Exception e)
		{
		}
		return this;
	}

	/**
	 * 设置sslSocketFactory
	 * @param sslSocketFactory SslSocketFactory
	 * @return HttpUtil
	 */
	public HttpUtil setSslSocketFactory(SSLSocketFactory sslSocketFactory)
	{
		this.sslSocketFactory = sslSocketFactory;
		return this;
	}

	/**
	 * 设置useCaches
	 * @param usecaches boolean
	 * @return HttpUtil
	 */
	public HttpUtil setUseCaches(boolean usecaches)
	{
		this.http.setUseCaches(usecaches);
		return this;
	}

	/**
	 * 设置userAgent
	 * @param userAgent String
	 * @return HttpUtil
	 */
	public HttpUtil setUserAgent(String userAgent)
	{
		this.userAgent = userAgent;
		if(http != null)
		{
			this.http.setRequestProperty("User-Agent", userAgent);
		}
		return this;
	}

	/**
	 * 创建新的http(s)请求，重置除cookie外的所有设置
	 * @param url url地址请求
	 * @return HttpUtil
	 */
	public HttpUtil create(String url)
	{
		return create(url, true);
	}

	/**
	 * 创建新的http(s)请求，重置除cookie、connectTimeout、readTimeout、userAgent外的所有设置
	 * @param url url地址请求
	 * @param isHostnameVerifier 是否不确认主机名
	 * @return HttpUtil
	 */
	public HttpUtil create(String url, boolean isHostnameVerifier)
	{
		this.clearForm();
		URL c;
		try
		{
			c = new URL(url);
			isHttps = c.getProtocol().toLowerCase().equals("https");
			this.http = (HttpURLConnection) c.openConnection();
			if(isHttps)
			{
				HttpsURLConnection https = (HttpsURLConnection) this.http;
				if(this.sslSocketFactory != null)
				{
					https.setSSLSocketFactory(this.sslSocketFactory);
				}
				if(isHostnameVerifier)
				{
					https.setHostnameVerifier(HttpCommon.HV);// 不进行主机名确认
				}
			}
			this.http.setDoInput(true);
			this.http.setDoOutput(false);
			this.http.setConnectTimeout(connectTimeout);
			this.http.setReadTimeout(readTimeout);
			this.http.setRequestProperty("User-Agent", userAgent);
			this.http.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
			this.http.setRequestProperty("Accept-Charset", "utf-8");
			this.http.setRequestMethod("GET");
		}
		catch(Exception e)
		{
		}
		return this;
	}

	/**
	 * 连接并返回网页文本
	 * @return 连接失败返回null
	 */
	public String connect()
	{
		return connect("UTF-8");
	}

	/**
	 * 连接并返回网页文本
	 * @param charsetName 对封装的表单、获取的网页内容进行的编码设置
	 * @return 连接失败返回null
	 */
	public String connect(String charsetName)
	{
		String result = null;
		try
		{
			if(this.cookies.size() > 0)
			{
				String _c = HttpCommon.parse(HttpCommon.getHttpCookies(this.cookies, isHttps()), "; ");
				http.setRequestProperty("Cookie", _c);
			}
			byte[] arr = null;
			if(this.form.size() > 0)
			{
				String data = HttpCommon.format(form, charsetName);
				arr = data.getBytes("ISO-8859-1");
			}
			else if(databody != null)
			{
				arr = databody;
			}
			if(arr != null)
			{
				this.http.setDoOutput(true);
				this.http.setUseCaches(false);
				if(this.http.getRequestMethod().toUpperCase().equals("GET"))// DELETE, PUT, POST
				{
					this.http.setRequestMethod("POST");
				}
				// this.http.setRequestProperty("Content-Length", String.valueOf(data.length()));
				DataOutputStream out = new DataOutputStream(this.http.getOutputStream());
				out.write(arr, 0, arr.length);
				// out.writeBytes(data);
				out.flush();
				out.close();
			}
			this.http.connect();
			int _responseCode = http.getResponseCode();// 设置http返回状态200（ok）还是403
			BufferedReader in = null;
			if(_responseCode >= 200 && _responseCode < 300)
			{
				Date date = new Date();
				List<Cookie> list = HttpCommon.getHttpCookies(http);
				for(Cookie m : list)
				{
					if(m.getExpiryDate() == null)
					{
						this.addCookie(m.getName(), m.getValue());// 会话cookie
					}
					else
					{
						if(!m.isExpired(date))
						{
							this.addCookie(m.getName(), m.getValue());
						}
					}
				}
				in = new BufferedReader(new InputStreamReader(http.getInputStream(), charsetName));
				String temp = in.readLine();
				while(temp != null)
				{
					if(result != null)
					{
						result += temp;
					}
					else
					{
						result = temp;
					}
					temp = in.readLine();
				}
				in.close();
			}
		}
		catch(Exception e)
		{
		}
		try
		{
			http.disconnect();
		}
		catch(Exception e)
		{
		}
		return result;
	}

	/**
	 * 连接并返回网页流
	 * @return 连接失败返回null
	 */
	public InputStream connectStream()
	{
		return connectStream("UTF-8");
	}

	/**
	 * 连接并返回网页流
	 * @param charsetName 对封装的表单、获取的网页内容进行的编码设置
	 * @return 连接失败返回null
	 */
	public InputStream connectStream(String charsetName)
	{
		try
		{
			if(this.cookies.size() > 0)
			{
				String _c = HttpCommon.parse(HttpCommon.getHttpCookies(this.cookies, isHttps()), "; ");
				http.setRequestProperty("Cookie", _c);
			}
			byte[] arr = null;
			if(this.form.size() > 0)
			{
				String data = HttpCommon.format(form, charsetName);
				arr = data.getBytes("ISO-8859-1");
			}
			else if(databody != null)
			{
				arr = databody;
			}
			if(arr != null)
			{
				this.http.setDoOutput(true);
				this.http.setUseCaches(false);
				if(this.http.getRequestMethod().toUpperCase().equals("GET"))// DELETE, PUT, POST
				{
					this.http.setRequestMethod("POST");
				}
				// this.http.setRequestProperty("Content-Length", String.valueOf(data.length()));
				DataOutputStream out = new DataOutputStream(this.http.getOutputStream());
				out.write(arr, 0, arr.length);
				// out.writeBytes(data);
				out.flush();
				out.close();
			}
			this.http.connect();
			int _responseCode = http.getResponseCode();// 设置http返回状态200（ok）还是403
			if(_responseCode >= 200 && _responseCode < 300)
			{
				Date date = new Date();
				List<Cookie> list = HttpCommon.getHttpCookies(http);
				for(Cookie m : list)
				{
					if(m.getExpiryDate() == null)
					{
						this.addCookie(m.getName(), m.getValue());// 会话cookie
					}
					else
					{
						if(!m.isExpired(date))
						{
							this.addCookie(m.getName(), m.getValue());
						}
					}
				}
				return http.getInputStream();
			}
		}
		catch(Exception e)
		{
		}
		try
		{
			http.disconnect();
		}
		catch(Exception e)
		{
		}
		return null;
	}
	
	public void initSocketFactoryForSSL()
	{
		this.sslSocketFactory = HttpCommon.getSocketFactoryForSSL();
	}

	public void initSocketFactoryForTLS()
	{
		this.sslSocketFactory = HttpCommon.getSocketFactoryForTLS();
	}
	
	// post的数据流，与Form数据冲突
	private byte[] databody = null;

	/**
	 * 设置数据流，优先使用form值
	 * @param arr byte[]
	 * @return HttpUtil
	 */
	public HttpUtil setDataBody(byte[] arr)
	{
		databody = arr;
		return this;
	}
	
	// 表单项
	private List<NameValue> form = new ArrayList<NameValue>();

	/**
	 * 清除已清加的表单项
	 * @return HttpUtil
	 */
	public HttpUtil clearForm()
	{
		form.clear();
		return this;
	}

	/**
	 * 添加表单项
	 * @param name String
	 * @param value String
	 * @return HttpUtil
	 */
	public HttpUtil addForm(String name, String value)
	{
		form.add(new NameValue(name, value));
		return this;
	}

	/**
	 * 批量添加表单项
	 * @param array NameValue[]
	 * @return HttpUtil
	 */
	public HttpUtil addForms(NameValue[] array)
	{
		for(NameValue c : array)
		{
			form.add(c);
		}
		return this;
	}
	// cookie
	private List<Cookie> cookies = new ArrayList<Cookie>();

	/**
	 * 清除已清加的cookie
	 * @return HttpUtil
	 */
	public HttpUtil clearCookies()
	{
		cookies.clear();
		return this;
	}

	/**
	 * 添加cookie
	 * @param name String
	 * @param value String
	 * @return HttpUtil
	 */
	public HttpUtil addCookie(String name, String value)
	{
		cookies.add(new Cookie(name, value));
		return this;
	}

	/**
	 * 批量添加cookie
	 * @param array Cookie[]
	 * @return HttpUtil
	 */
	public HttpUtil addCookies(Cookie[] array)
	{
		for(Cookie c : array)
		{
			cookies.add(c);
		}
		return this;
	}

	/**
	 * 复制cookie
	 * @param onlySessionCookie true：仅复制会话cookie false：复制非会话cookie null:全部cookie
	 * @return List&lt;Cookie&gt;
	 */
	public List<Cookie> getCloneCookies(Boolean onlySessionCookie)
	{
		List<Cookie> lists = HttpCommon.getHttpCookies(this.cookies, true);
		List<Cookie> list = new ArrayList<Cookie>();
		if(onlySessionCookie == null)
		{
			for(Cookie m : lists)
			{
				list.add(m.clone());
			}
		}
		else if(onlySessionCookie)
		{
			for(Cookie m : lists)
			{
				if(m.getExpiryDate() == null)
				{
					list.add(m.clone());
				}
			}
		}
		else
		{
			for(Cookie m : lists)
			{
				if(m.getExpiryDate() != null)
				{
					list.add(m.clone());
				}
			}
		}
		return list;
	}
}

File: DsWorkNet/Dswork.Http/HttpUtil.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Text;

namespace Dswork.Http
{
	/// <summary>
	/// 扩展Request的功能
	/// </summary>
	public class HttpUtil
	{
		private HttpWebRequest http;
		private Boolean isHttps = false;
		private int connectTimeout = 10000;
		private int readTimeout = 30000;
		private String userAgent = "Mozilla/5.0 (compatible; MSIE 11; Windows NT 6.1; Win64; x64;)";// Gecko/20150123

		/// <summary>
		/// 返回当前是否https请求
		/// </summary>
		public Boolean Https
		{
			get { return isHttps; }
		}

		/// <summary>
		/// 设置超时时间毫秒
		/// </summary>
		/// <param name="connectTimeout">int</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil SetConnectTimeout(int connectTimeout)
		{
			this.connectTimeout = connectTimeout;
			if (http != null)
			{
				try
				{
					this.http.Timeout = connectTimeout;
				}
				catch
				{
				}
			}
			return this;
		}

		/// <summary>
		/// 设置contentType
		/// </summary>
		/// <param name="contentType">String</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil SetContentType(String contentType)
		{
			if (http != null)
			{
				this.http.ContentType = contentType;
			}
			return this;
		}

		/// <summary>
		/// 设置读取超时时间毫秒
		/// </summary>
		/// <param name="readTimeout">int</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil SetReadTimeout(int readTimeout)
		{
			this.readTimeout = readTimeout;
			if (http != null)
			{
				try
				{
					this.http.ReadWriteTimeout = readTimeout;
				}
				catch
				{
				}
			}
			return this;
		}

		/// <summary>
		/// 设置requestMethod
		/// </summary>
		/// <param name="requestMethod">String</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil SetRequestMethod(String requestMethod)
		{
			if (http != null)
			{
				this.http.Method = requestMethod.ToUpper();
			}
			return this;
		}

		/// <summary>
		/// 设置requestProperty
		/// </summary>
		/// <param name="key">String</param>
		/// <param name="value">String</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil SetRequestProperty(String key, String value)
		{
			if (http != null)
			{
				try
				{
					this.http.Headers[key] = value;
				}
				catch
				{
				}
			}
			return this;
		}

		/// <summary>
		/// 设置userAgent
		/// </summary>
		/// <param name="userAgent">String</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil SetUserAgent(String userAgent)
		{
			this.userAgent = userAgent;
			if (http != null)
			{
				this.http.Accept = "";
				this.http.UserAgent = userAgent;
			}
			return this;
		}

		/// <summary>
		/// 创建新的http(s)请求，重置除cookie外的所有设置
		/// </summary>
		/// <param name="url">url地址请求</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil Create(String urlpath)
		{
			return Create(urlpath, true);
		}

		/// <summary>
		/// 创建新的http(s)请求，重置除cookie外的所有设置
		/// </summary>
		/// <param name="url">url地址请求</param>
		/// <param name="isHostnameVerifier">是否不确认主机名</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil Create(String url, Boolean isHostnameVerifier)
		{
			this.ClearForm();
			try
			{
				http = WebRequest.Create(url) as HttpWebRequest;
				isHttps = url.StartsWith("https", StringComparison.OrdinalIgnoreCase);
				if (isHostnameVerifier && isHttps)
				{
					HttpCommon.HostnameVerifier();
				}
				this.http.Timeout = connectTimeout;
				this.http.ReadWriteTimeout = readTimeout;
				this.http.UserAgent = userAgent;
				this.http.ContentType = "application/x-www-form-urlencoded";
				this.http.Headers["Accept-Charset"] = "utf-8";
				this.http.Method = "GET";
			}
			catch
			{
			}
			return this;
		}

		public String Connect()
		{
			return Connect("UTF-8");
		}

		public String Connect(String charsetName)
		{
			String result = null;
			try
			{
				DateTime dt = new DateTime();
				this.http.CookieContainer = new CookieContainer();
				if (this.cookies.Count > 0)
				{
					/*
					List<Cookie> list = HttpCommon.GetHttpCookies(this.cookies, Https);
					foreach (Cookie m in list)
					{
						if (m.ExpiryDate == null || (m.ExpiryDate != null && !m.IsExpired(dt)))
						{
							System.Net.Cookie x = new System.Net.Cookie();
							x.Name = m.Name;
							x.Value = m.Value;
							x.Secure = m.Secure;
							x.Path = m.Path;
							x.Domain = m.Domain;
							x.Expires = m.ExpiryDate;
							x.HttpOnly = m.HttpOnly;
							this.http.CookieContainer.Add(x);
						}
					}
					*/
					String _c = HttpCommon.parse(HttpCommon.GetHttpCookies(this.cookies, Https), "; ");
					this.http.CookieContainer.SetCookies(http.RequestUri, _c);
				}
				if (this.form.Count > 0)
				{
					String data = HttpCommon.format(form, charsetName);
					if (this.http.Method.ToUpper() == "GET")
					{
						this.http.Method = "POST";
					}
					Encoding enc = Encoding.GetEncoding(28591);//28591对应iso-8859-1
					using (Stream stream = this.http.GetRequestStream())
					{
						stream.Write(enc.GetBytes(data), 0, data.Length);
					}
				}
				HttpWebResponse res = http.GetResponse() as HttpWebResponse;
				if (res.StatusCode == HttpStatusCode.OK)
				{
					List<Cookie> list = HttpCommon.GetHttpCookies(res.Cookies);
					foreach (Cookie m in list)
					{
						if (m.ExpiryDate == null)
						{
							this.AddCookie(m.Name, m.Value);// 会话cookie
						}
						else
						{
							if (!m.IsExpired(dt))
							{
								this.AddCookie(m.Name, m.Value);
							}
						}
					}
					Stream stream = res.GetResponseStream();
					Encoding ee = charsetName.ToLower().Equals("utf-8") ? new UTF8Encoding(false) : Encoding.GetEncoding(charsetName);
					StreamReader streamReader = new StreamReader(stream, ee);
					result = streamReader.ReadToEnd();
					streamReader.Close();
					stream.Close();
					res.Close();
				}
			}
			catch(Exception ex)
			{
				Console.WriteLine(ex.Message);
			}
			return result;
		}

		// 表单项
		private List<NameValue> form = new List<NameValue>();
		/// <summary>
		/// 清除已清加的表单项
		/// </summary>
		///// <returns>HttpUtil</returns>
		public HttpUtil ClearForm()
		{
			form.Clear();
			return this;
		}

		/// <summary>
		/// 添加表单项
		/// </summary>
		/// <param name="name">name</param>
		/// <param name="value">value</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil AddForm(String name, String value)
		{
			form.Add(new NameValue(name, value));
			return this;
		}

		/// <summary>
		/// 批量添加表单项
		/// </summary>
		/// <param name="name">name</param>
		/// <param name="value">value</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil AddForms(NameValue[] array)
		{
			foreach (NameValue c in array)
			{
				form.Add(c);
			}
			return this;
		}

		// cookie
		private List<Cookie> cookies = new List<Cookie>();

		/// <summary>
		/// 清除已清加的cookie
		/// </summary>
		/// <returns>HttpUtil</returns>
		public HttpUtil ClearCookies()
		{
			cookies.Clear();
			return this;
		}

		/// <summary>
		/// 添加cookie
		/// </summary>
		/// <param name="name">name</param>
		/// <param name="value">value</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil AddCookie(String name, String value)
		{
			cookies.Add(new Cookie(name, value));
			return this;
		}

		/// <summary>
		/// 批量添加cookie
		/// </summary>
		/// <param name="name">name</param>
		/// <param name="value">value</param>
		/// <returns>HttpUtil</returns>
		public HttpUtil AddCookies(Cookie[] array)
		{
			foreach (Cookie c in array)
			{
				cookies.Add(c);
			}
			return this;
		}

		/// <summary>
		/// 复制cookie
		/// </summary>
		/// <param name="onlySessionCookie">true：仅复制会话cookie false：复制非会话cookie null:全部cookie</param>
		/// <returns>List<Cookie></returns>
		public List<Cookie> GetCloneCookies(Boolean onlySessionCookie)
		{
			List<Cookie> lists = HttpCommon.GetHttpCookies(this.cookies, true);
			List<Cookie> list = new List<Cookie>();
			if (onlySessionCookie)
			{
				foreach (Cookie m in lists)
				{
					if (m.ExpiryDate == null)
					{
						list.Add(m.Clone());
					}
				}
			}
			else
			{
				foreach (Cookie m in lists)
				{
					if (m.ExpiryDate != null)
					{
						list.Add(m.Clone());
					}
				}
			}
			return list;
		}
	}
}
